pystac/item.py

```diff
@@ -49,33 +49,49 @@ class Item(STACObject):
             belongs to.
         extra_fields : Extra fields that are part of the top-level JSON
             properties of the Item.
-
-    Attributes:
-        id : Provider identifier. Unique within the STAC.
-        geometry : Defines the full footprint of the asset represented by this
-            item, formatted according to
-            `RFC 7946, section 3.1 (GeoJSON) <https://tools.ietf.org/html/rfc7946>`_.
-        bbox :  Bounding Box of the asset represented by this item
-            using either 2D or 3D geometries. The length of the array is 2*n where n
-            is the number of dimensions. Could also be None in the case of a null
-            geometry.
-        datetime : Datetime associated with this item. If None,
-            the start_datetime and end_datetime in the common_metadata
-            will supply the datetime range of the Item.
-        properties : A dictionary of additional metadata for the item.
-        stac_extensions : Optional list of extensions the Item
-            implements.
-        collection : Collection that this item is a part of.
-        links : A list of :class:`~pystac.Link` objects representing
-            all links associated with this STACObject.
-        assets : Dictionary of asset objects that can be downloaded,
-            each with a unique key.
-        collection_id : The Collection ID that this item belongs to, if
-            any.
-        extra_fields : Extra fields that are part of the top-level JSON
-            properties of the Item.
     """
 
+    assets: Dict[str, Asset]
+    """Dictionary of :class:`~pystac.Asset` objects, each with a unique key."""
+
+    bbox: Optional[List[float]]
+    """Bounding Box of the asset represented by this item using either 2D or 3D
+    geometries. The length of the array is 2*n where n is the number of dimensions.
+    Could also be None in the case of a null geometry."""
+
+    collection: Optional[Collection]
+    """Collection that this item is a part of, if any."""
+
+    collection_id: Optional[str]
+    """The Collection ID that this item belongs to, if any."""
+
+    datetime: Optional[Datetime]
+    """Datetime associated with this item. If ``None``, then
+    :attr:`~pystac.CommonMetadata.start_datetime` and
+    :attr:`~pystac.CommonMetadata.end_datetime` in :attr:`~pystac.Item.common_metadata`
+    will supply the datetime range of the Item."""
+
+    extra_fields: Dict[str, Any]
+    """Extra fields that are part of the top-level JSON fields the Item."""
+
+    geometry: Optional[Dict[str, Any]]
+    """Defines the full footprint of the asset represented by this item, formatted
+    according to `RFC 7946, section 3.1 (GeoJSON)
+    <https://tools.ietf.org/html/rfc7946>`_."""
```

Comment:
Since this is referring to a known standard, would it be possible to lock it down more than `Any`? I'm not familiar enough with GeoJSON to tell whether that'd be a lot of work, though.

Reply:
The full type signature is a mess: something like `Dict[str, Union[List[float], List[Union[List[float], float]]]]` - I'm not sure that goes deep enough for the unions. IMO json dicts should generally be classified as `Dict[str, Any]` including geojson.

Reply:
Yeah, the [Rust `geojson` library](https://docs.rs/geojson/0.22.2/geojson/enum.Value.html) has some interesting recursive typing to handle GeoJSON that we might be able to borrow, but I agree that it's probably not worth it. It seems like it would only be a useful check if users are using a similar level of specificity when typing their inputs. I think it's more likely that users will be using `Dict[str, Any]` to type any GeoJSON they use here, and making this too specific might lead to typing errors for them.

Reply:
@lossyrob That type isn't so bad - it'll give programmers a lot more information than `Dict[str, Any]`. But if the type ends up being recursive we might as well give up, because Mypy doesn't support recursive types yet.

Reply:
Actually, to accommodate a basic MultiPolygon we would need `Dict[str, Union[str, List[float], List[List[float]], List[List[List[float]]], List[List[List[List[float]]]]]]`. I haven't tried it out, but I think we would need a recursive definition to accommodate GeometryCollections since GeometryCollections have a `"geometry"` field that is an array of GeoJSON Geometries. The GeoJSON spec also does not prohibit other fields in addition to `"type"`, `"coordinates"`, and `"geometry"`, which means we would probably need something more like `Dict[str, Union[str, Any, List[float], List[List[float]], List[List[List[float]]], List[List[List[List[float]]]]]]`, which seems like it is starting to be more confusing than helpful.
My feeling is that it is more helpful to link to the spec and if people want a more rigorous check they can use the validation features.